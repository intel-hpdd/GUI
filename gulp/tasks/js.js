//
// INTEL CONFIDENTIAL
//
// Copyright 2013-2016 Intel Corporation All Rights Reserved.
//
// The source code contained or described herein and all documents related
// to the source code ("Material") are owned by Intel Corporation or its
// suppliers or licensors. Title to the Material remains with Intel Corporation
// or its suppliers and licensors. The Material contains trade secrets and
// proprietary and confidential information of Intel or its suppliers and
// licensors. The Material is protected by worldwide copyright and trade secret
// laws and treaty provisions. No part of the Material may be used, copied,
// reproduced, modified, published, uploaded, posted, transmitted, distributed,
// or disclosed in any way without Intel's prior express written permission.
//
// No license under any patent, copyright, trade secret or other intellectual
// property right is granted to or conferred upon you by disclosure or delivery
// of the Materials, either expressly, by implication, inducement, estoppel or
// otherwise. Any license under such intellectual property rights must be
// express and approved by Intel in writing.

'use strict';

var paths = require('../paths.json');
var gulp = require('gulp');
var babel = require('gulp-babel');
var sourcemaps = require('gulp-sourcemaps');
var destDir = require('../dest-dir');
var relativeSourcemapsSource = require('gulp-relative-sourcemaps-source');

function jsDeps (fn) {
  return gulp.src(paths.js.deps, {
    since: gulp.lastRun(fn),
    base: '.'
  });
}

exports.jsDepsDev = function jsDepsDev () {
  return jsDeps(jsDepsDev)
  .pipe(gulp.dest('./dest'))
  .pipe(gulp.symlink('static/chroma_ui', { cwd: destDir }));
};

exports.jsDepsProd = function jsDepsProd () {
  return jsDeps(jsDepsProd)
  .pipe(gulp.dest('./dest'));
};

function socketWorker (fn) {
  return gulp.src(paths.js.socketWorker, {
    since: gulp.lastRun(fn)
  })
  .pipe(sourcemaps.init({ loadMaps: true }))
  .pipe(sourcemaps.write('.'));
}

exports.socketWorkerDev = function socketWorkerDev () {
  return socketWorker(socketWorkerDev)
  .pipe(gulp.dest('./dest'))
  .pipe(gulp.symlink('static/chroma_ui', { cwd: destDir }));
};

exports.socketWorkerProd = function socketWorkerProd () {
  return socketWorker(socketWorkerProd)
  .pipe(gulp.dest('./dest'))
  .pipe(gulp.dest('./dist'));
};

function systemConfig (fn) {
  return gulp.src(paths.js.config, {
    since: gulp.lastRun(fn),
    base: '.'
  });
}

exports.systemConfigDev = function systemConfigDev () {
  return systemConfig(systemConfigDev)
  .pipe(gulp.dest('./dest'))
  .pipe(gulp.symlink('static/chroma_ui', { cwd: destDir }));
};

var babelDev = babel.bind(null, {
  presets: [],
  babelrc: false,
  plugins: [
    'check-es2015-constants',
    'transform-flow-strip-types',
    'transform-class-properties',
    'transform-object-rest-spread',
    'transform-async-to-generator',
    'transform-es2015-classes',
    'transform-es2015-modules-systemjs'
  ]
});

var babelProd = babel.bind(null, {
  presets: ['es2015'],
  babelrc: false,
  plugins: [
    'transform-flow-strip-types',
    'transform-class-properties',
    'transform-object-rest-spread',
    [
      'angularjs-annotate',
      {
        explicitOnly: true
      }
    ],
    'transform-es2015-modules-systemjs'
  ]
});

function jsSource (fn) {
  return gulp.src(paths.js.source, {
    since: gulp.lastRun(fn),
    base: '.'
  })
  .pipe(sourcemaps.init());
}

exports.jsSourceDev = function jsSourceDev () {
  return jsSource(jsSourceDev)
  .pipe(babelDev())
  .pipe(sourcemaps.write({ sourceRoot: '/static/chroma_ui' }))
  .pipe(gulp.dest('./dest'))
  .pipe(gulp.symlink('static/chroma_ui', { cwd: destDir }));
};

exports.jsSourceProd = function jsSourceProd () {
  return jsSource(jsSourceProd)
  .pipe(babelProd())
  .pipe(relativeSourcemapsSource({dest: 'dest'}))
  .pipe(sourcemaps.write({ sourceRoot: '' }))
  .pipe(gulp.dest('./dest'));
};

exports.jsTest = function jsTest () {
  return gulp.src(paths.js.tests, {
    since: gulp.lastRun(jsTest),
    base: '.'
  })
  .pipe(sourcemaps.init())
  .pipe(babelDev())
  .pipe(sourcemaps.write({ sourceRoot: '' }))
  .pipe(gulp.dest('./dest'));
};

exports.jsTestFixtures = function jsTestFixtures () {
  return gulp.src(paths.js.testFixtures, {
    since: gulp.lastRun(jsTestFixtures),
    base: '.'
  })
  .pipe(gulp.dest('./dest'));
};

exports.jsTestDeps = function jsTestDeps () {
  return gulp.src(paths.js.testDeps, {
    since: gulp.lastRun(jsTestDeps),
    base: '.'
  })
  .pipe(gulp.dest('./dest'));
};
