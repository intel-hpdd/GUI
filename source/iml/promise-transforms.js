// @flow

//
// INTEL CONFIDENTIAL
//
// Copyright 2013-2016 Intel Corporation All Rights Reserved.
//
// The source code contained or described herein and all documents related
// to the source code ("Material") are owned by Intel Corporation or its
// suppliers or licensors. Title to the Material remains with Intel Corporation
// or its suppliers and licensors. The Material contains trade secrets and
// proprietary and confidential information of Intel or its suppliers and
// licensors. The Material is protected by worldwide copyright and trade secret
// laws and treaty provisions. No part of the Material may be used, copied,
// reproduced, modified, published, uploaded, posted, transmitted, distributed,
// or disclosed in any way without Intel's prior express written permission.
//
// No license under any patent, copyright, trade secret or other intellectual
// property right is granted to or conferred upon you by disclosure or delivery
// of the Materials, either expressly, by implication, inducement, estoppel or
// otherwise. Any license under such intellectual property rights must be
// express and approved by Intel in writing.

import * as fp from 'intel-fp';

import type {
  HighlandStreamT
} from 'highland';

export const resolveStream = (stream:HighlandStreamT<mixed>) => {
  return new Promise((resolve) => {
    stream.pull((error, x) => {
      if (error)
        x = {
          __HighlandStreamError__: true,
          error
        };

      const s2 = stream.tap(fp.noop);
      s2.write(x);

      // $FlowIgnore: flow does not recognize this monkey-patching.
      s2.destroy = stream.destroy.bind(stream);

      resolve(s2);
    });
  });
};


export function streamToPromise (s:HighlandStreamT<Object>):Promise<Object> {
  return new Promise ((resolve:Function, reject:Function) => {
    return s
    .pull((err:Error, x:Object) => {
      if (err)
        reject(err);
      else
        resolve(x);
    });
  })
  .then(fp.tap(s.destroy.bind(s)));
}

export const extendParentData = fp.curry(2,
  (parent:Object, x:Object):Object =>
    ({
      ...x,
      parent
    })
);
