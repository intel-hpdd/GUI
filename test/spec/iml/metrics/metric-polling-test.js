import { streamToPromise } from "../../../../source/iml/promise-transforms.js";

describe("polling metrics from filesystem", () => {
  let data, metric$, metricPoll, mockGlobal, mockEnvironment;

  beforeEach(() => {
    data = {
      "1": [
        {
          data: {
            client_count: 0,
            filesfree: 3275995,
            filestotal: 3276800,
            kbytesavail: 3194296,
            kbytesfree: 5544860,
            kbytestotal: 7736392,
            num_exports: 2,
            stats_close: 3,
            stats_create: 4,
            stats_get_info: 3,
            stats_getattr: 9,
            stats_getxattr: 2,
            stats_ldlm_ibits_enqueue: 14,
            stats_mds_connect: 1100264,
            stats_mds_disconnect: 15813,
            stats_mds_get_root: 30,
            stats_mds_getattr: 140,
            stats_mds_reint_open: 3,
            stats_mds_statfs: 180,
            stats_mknod: 2,
            stats_obd_ping: 3701378,
            stats_open: 3,
            stats_read_bytes: 1073741824,
            stats_read_iops: 256,
            stats_req_active: 54928,
            stats_req_qdepth: 54928,
            stats_req_timeout: 73756,
            stats_req_waittime: 13469486,
            stats_reqbuf_avail: 9355970,
            stats_set_info: 12,
            stats_statfs: 127656,
            stats_write_bytes: 2147483648,
            stats_write_iops: 514,
            tot_dirty: 0,
            tot_granted: 558208,
            tot_pending: 0
          },
          ts: "2019-11-05T20:08:50+00:00"
        },
        {
          data: {
            client_count: 0,
            filesfree: 3275995,
            filestotal: 3276800,
            kbytesavail: 3194296,
            kbytesfree: 5544860,
            kbytestotal: 7736392,
            num_exports: 2,
            stats_close: 3,
            stats_create: 4,
            stats_get_info: 3,
            stats_getattr: 9,
            stats_getxattr: 2,
            stats_ldlm_ibits_enqueue: 14,
            stats_mds_connect: 1100264,
            stats_mds_disconnect: 15813,
            stats_mds_get_root: 30,
            stats_mds_getattr: 140,
            stats_mds_reint_open: 3,
            stats_mds_statfs: 180,
            stats_mknod: 2,
            stats_obd_ping: 3701378,
            stats_open: 3,
            stats_read_bytes: 1073741824,
            stats_read_iops: 256,
            stats_req_active: 54928,
            stats_req_qdepth: 54928,
            stats_req_timeout: 73756,
            stats_req_waittime: 13469486,
            stats_reqbuf_avail: 9355970,
            stats_set_info: 12,
            stats_statfs: 127656,
            stats_write_bytes: 2147483648,
            stats_write_iops: 514,
            tot_dirty: 0,
            tot_granted: 558208,
            tot_pending: 0
          },
          ts: "2019-11-05T20:09:00+00:00"
        }
      ],
      "2": [
        {
          data: {
            client_count: 2,
            filesfree: 5897048,
            filestotal: 5898240,
            kbytesavail: 5291184,
            kbytesfree: 9435328,
            kbytestotal: 9556360,
            num_exports: 4,
            stats_close: 0,
            stats_create: 8,
            stats_get_info: 4,
            stats_getattr: 3,
            stats_getxattr: 0,
            stats_ldlm_ibits_enqueue: 15,
            stats_mds_connect: 2125720,
            stats_mds_disconnect: 16098,
            stats_mds_get_root: 90,
            stats_mds_getattr: 264,
            stats_mds_reint_open: 3,
            stats_mds_statfs: 246,
            stats_mknod: 0,
            stats_obd_ping: 3707352,
            stats_open: 0,
            stats_read_bytes: 0,
            stats_read_iops: 0,
            stats_req_active: 55084,
            stats_req_qdepth: 55084,
            stats_req_timeout: 73930,
            stats_req_waittime: 13489856,
            stats_reqbuf_avail: 9380610,
            stats_set_info: 0,
            stats_statfs: 1035,
            stats_write_bytes: 0,
            stats_write_iops: 0,
            tot_dirty: 0,
            tot_granted: 556416,
            tot_pending: 0
          },
          ts: "2019-11-05T20:08:50+00:00"
        },
        {
          data: {
            client_count: 2,
            filesfree: 5897048,
            filestotal: 5898240,
            kbytesavail: 5291184,
            kbytesfree: 9435328,
            kbytestotal: 9556360,
            num_exports: 4,
            stats_close: 0,
            stats_create: 8,
            stats_get_info: 4,
            stats_getattr: 3,
            stats_getxattr: 0,
            stats_ldlm_ibits_enqueue: 15,
            stats_mds_connect: 2125720,
            stats_mds_disconnect: 16098,
            stats_mds_get_root: 90,
            stats_mds_getattr: 264,
            stats_mds_reint_open: 3,
            stats_mds_statfs: 246,
            stats_mknod: 0,
            stats_obd_ping: 3707352,
            stats_open: 0,
            stats_read_bytes: 0,
            stats_read_iops: 0,
            stats_req_active: 55084,
            stats_req_qdepth: 55084,
            stats_req_timeout: 73930,
            stats_req_waittime: 13489856,
            stats_reqbuf_avail: 9380610,
            stats_set_info: 0,
            stats_statfs: 1035,
            stats_write_bytes: 0,
            stats_write_iops: 0,
            tot_dirty: 0,
            tot_granted: 556416,
            tot_pending: 0
          },
          ts: "2019-11-05T20:09:00+00:00"
        }
      ]
    };

    mockGlobal = {
      fetch: jest.fn()
    };
    jest.mock("../../../../source/iml/global.js", () => mockGlobal);

    mockEnvironment = {
      API: "https://localhost:443/api/"
    };
    jest.mock("../../../../source/iml/environment.js", () => mockEnvironment);

    metricPoll = require("../../../../source/iml/metrics/metric-polling.js").metricPoll;
  });

  describe("with data", () => {
    beforeEach(() => {
      mockGlobal.fetch.mockImplementation(() => ({
        json: () => data
      }));
    });

    it("should extract the metric data", async () => {
      metric$ = await metricPoll();
      const transformedData = await streamToPromise(metric$);
      expect(transformedData).toMatchSnapshot();
    });
  });

  describe("without data", () => {
    beforeEach(() => {
      mockGlobal.fetch.mockImplementation(() => ({
        json: () => ({
          "1": [{ data: {}, ts: "" }],
          "2": [{ data: { ...data["2"][0].data }, ts: "" }]
        })
      }));
    });

    it("should extract the metric data", async () => {
      metric$ = await metricPoll();
      const transformedData = await streamToPromise(metric$);
      expect(transformedData).toMatchSnapshot();
    });
  });

  describe("with missing data", () => {
    beforeEach(() => {
      mockGlobal.fetch.mockImplementation(() => ({
        json: () => ({
          "1": [
            {
              data: {
                client_count: 1,
                filesfree: 3275995,
                kbytesavail: 3194296,
                kbytesfree: 5544860
              },
              ts: ""
            }
          ],
          "2": [{ data: { ...data["2"][0].data }, ts: "" }]
        })
      }));
    });

    it("should extract the metric data", async () => {
      metric$ = await metricPoll();
      const transformedData = await streamToPromise(metric$);
      expect(transformedData).toMatchSnapshot();
    });
  });

  describe("with a single entry for the filesystem", () => {
    beforeEach(() => {
      mockGlobal.fetch.mockImplementation(() => ({
        json: () => ({
          "1": [{ ...data["1"][0] }]
        })
      }));
    });

    it("should extract the metric data", async () => {
      metric$ = await metricPoll();
      const transformedData = await streamToPromise(metric$);
      expect(transformedData).toMatchSnapshot();
    });
  });
});
