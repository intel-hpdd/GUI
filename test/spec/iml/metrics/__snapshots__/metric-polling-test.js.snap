// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`polling metrics from filesystem with a single entry for the filesystem should extract the metric data 1`] = `
Object {
  "1": Object {
    "bytes_free": 5677936640,
    "bytes_total": 7922065408,
    "client_count": 0,
    "files_free": 3354618880,
    "files_total": 3355443200,
  },
}
`;

exports[`polling metrics from filesystem with data should extract the metric data 1`] = `
Object {
  "1": Object {
    "bytes_free": 5677936640,
    "bytes_total": 7922065408,
    "client_count": 0,
    "files_free": 3354618880,
    "files_total": 3355443200,
  },
  "2": Object {
    "bytes_free": 9661775872,
    "bytes_total": 9785712640,
    "client_count": 2,
    "files_free": 6038577152,
    "files_total": 6039797760,
  },
}
`;

exports[`polling metrics from filesystem with missing data should extract the metric data 1`] = `
Object {
  "1": Object {
    "bytes_free": 5677936640,
    "bytes_total": undefined,
    "client_count": 1,
    "files_free": 3354618880,
    "files_total": undefined,
  },
  "2": Object {
    "bytes_free": 9661775872,
    "bytes_total": 9785712640,
    "client_count": 2,
    "files_free": 6038577152,
    "files_total": 6039797760,
  },
}
`;

exports[`polling metrics from filesystem without data should extract the metric data 1`] = `
Object {
  "1": Object {
    "bytes_free": undefined,
    "bytes_total": undefined,
    "client_count": undefined,
    "files_free": undefined,
    "files_total": undefined,
  },
  "2": Object {
    "bytes_free": 9661775872,
    "bytes_total": 9785712640,
    "client_count": 2,
    "files_free": 6038577152,
    "files_total": 6039797760,
  },
}
`;
